underfs/cephfs/src/main/java/alluxio/underfs/cephfs/CephFSUnderFileSystem.java

```diff
@@ -0,0 +1,792 @@
+import alluxio.underfs.BaseUnderFileSystem;
+import alluxio.underfs.UfsDirectoryStatus;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.UnderFileSystemConfiguration;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.underfs.options.FileLocationOptions;
+import alluxio.underfs.options.MkdirsOptions;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.UnderFileSystemUtils;
+
+import alluxio.util.io.PathUtils;
+
+import com.ceph.fs.CephFileAlreadyExistsException;
+import com.ceph.fs.CephMount;
+import com.ceph.fs.CephNotDirectoryException;
+import com.ceph.fs.CephStat;
+import com.ceph.fs.CephStatVFS;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import java.util.List;
+import java.util.Stack;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * CephFS {@link UnderFileSystem} implementation.
+ */
+@ThreadSafe
+public class CephFSUnderFileSystem extends BaseUnderFileSystem
+    implements AtomicFileOutputStreamCallback {
+  private static final Logger LOG = LoggerFactory.getLogger(CephFSUnderFileSystem.class);
+  private static final int MAX_TRY = 5;
+
+  private CephMount mMount;
+
+  /**
+   * Factory method to constructs a new CephFS {@link UnderFileSystem} instance.
+   *
+   * @param ufsUri the {@link AlluxioURI} for this UFS
+   * @param conf the configuration for Hadoop
+   * @return a new CephFS {@link UnderFileSystem} instance
+   */
+  public static CephFSUnderFileSystem createInstance(
+      AlluxioURI ufsUri, UnderFileSystemConfiguration conf) throws IOException {
+    LOG.info("CephFS URI: {}", ufsUri.toString());
+
+    /*
+     * Create mount with auth user id
+     */
+    String userId = conf.get(PropertyKey.UNDERFS_CEPHFS_AUTH_ID);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_AUTH_ID, userId);
+    if (userId != null && userId.isEmpty()) {
+      userId = null;
+    }
+    CephMount mount = new CephMount(userId);
+
+    /*
+     * Load a configuration file if specified
+     */
+    String configfile = conf.get(PropertyKey.UNDERFS_CEPHFS_CONF_FILE);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_CONF_FILE, configfile);
+    if (configfile != null && !configfile.isEmpty()) {
+      mount.conf_read_file(configfile);
+    }
+
+    /*
+     * Parse and set Ceph configuration options
+     */
+    String configopts = conf.get(PropertyKey.UNDERFS_CEPHFS_CONF_OPTS);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_CONF_OPTS, configopts);
+    if (configopts != null && !configopts.isEmpty()) {
+      String[] options = configopts.split(";");
+      for (String option : options) {
+        String[] keyval = option.split("=");
```

Comment:
As the value of key maybe `xxxxx==`, Can you use some more complex separator instead?

Reply:
After an offline discussion with @jhonxue , it can make sense to keep the separator to `=`, and we shouldn't put all other non-options config into opts config parameter.